Lesson_19/homework_19_1.py

```diff
@@ -0,0 +1,145 @@
+        try:
+            # Send request to NASA API to get photos data
+            response = requests.get(
+                self.url, self.params, timeout=self.timeout)
+            response.raise_for_status()  # Rase an exception if code is not 2xx
+
+            # Get JSON data from response
+            try:
+                data = response.json()
+                photos = data.get('photos')
+
+                if photos:
+                    self.download_photos(photos)
+                else:
+                    _log.info('No photos found.')
+
+            except json.JSONDecodeError as error:
+                error = 'Error serializing JSON: {error}'
+                _log.error(error)
+        except requests.exceptions.HTTPError as error:
+            error = f'HTTP Error: {error}'
+            _log.error(error)
+        except requests.exceptions.ConnectionError as error:
+            error = f'Connection Error: {error}'
+            _log.error(error)
+        except requests.exceptions.Timeout as error:
+            error = f'Timeout Error: {error}'
+            _log.error(error)
+        except requests.exceptions.RequestException as error:
+            error = f'Request failed: {error}'
+            _log.error(error)
+
+    def download_photos(self, photos):
+        """Download photos from their URLs and saves them locally.
+
+        Args:
+            photos (list): List of photo metadata dictionaries.
+        """
+        for index, photo in enumerate(photos, start=1):
+            img_url = photo.get('img_src')
+            if img_url:
+                self.save_image(img_url, index)
+            else:
+                message = f'Photo {index} has no image URL.'
+                _log.info(message)
+
+    def save_image(self, img_url, index):
+        """Save an image locally from its URL.
+
+        Args:
+            img_url (str): URL of the image to download.
+            index (int): Index for naming the file.
+        """
+        try:
+            # Send a request to download the image
+            img_response = requests.get(img_url, timeout=self.timeout)
+            img_response.raise_for_status()
+            with open(f'mars_photo{index}.jpg', 'wb') as img_file:
+                img_file.write(img_response.content)
+                message = f'Mars_photo{index}.jpg downloaded successfully.'
+                _log.info(message)
+        except requests.exceptions.HTTPError as error:
+            error = f'HTTP Error: {error}'
+            _log.error(error)
+        except requests.exceptions.ConnectionError as error:
+            error = f'Connection Error: {error}'
+            _log.error(error)
+        except requests.exceptions.Timeout as error:
+            error = f'Timeout Error: {error}'
+            _log.error(error)
+        except requests.exceptions.RequestException as error:
+            error = f'Request failed: {error}'
+            _log.error(error)
+
```

Comment:
In general code looks too heavy for such simple solution - try to optimize

Reply:
Fixed in commit 31dfb0b758979996cd1a5d4f706a0d589d4b24f9